colab_evaluation.py

```diff
@@ -0,0 +1,226 @@
+are essentially zero, and so only contribute to false negatives. It therefore represents the data in "tidy
+format" with one row per example-label pair and excludes label-example pairs below a defined threshold.
+"""
+
+import numpy as np
+import pandas as pd
+import sklearn
+import tqdm
+
+import inference
+import utils
+import evaluation
+
+
+def batch_inferences(iterator, batch_size):
+    """Yield batches of seq_ids and predictions matrices from an iterator."""
+    counter = 0
+    predictions = []
+    seq_ids = []
+    while True:
+        try:
+            inference = next(iterator)
+        except StopIteration:
+            if len(seq_ids) > 0:
+                yield seq_ids, np.vstack(predictions)
+            return
+        seq_ids.append(inference[0])
+        predictions.append(inference[1])
+        counter += 1
+        if counter == batch_size:
+            yield seq_ids, np.vstack(predictions)
+            predictions = []
+            seq_ids = []
+            counter = 0
+
+
+def batched_inferences_from_files(shard_names, batch_size=100):
+    """Iterate through TFRecord files of inferences and yield batches."""
+    for file_name in tqdm.tqdm(shard_names, position=0):
+        inference_iterator = inference.parse_shard(file_name)
+        batched_iterator = batch_inferences(inference_iterator, batch_size)
+        while True:
+            try:
+                yield next(batched_iterator)
+            except StopIteration:
+                break
+
+
+def batched_inferences_from_dir(shard_dir_path, batch_size=100):
+    """Iterate through directory of inference TFRecord files and yield batches."""
+    files_to_process = utils.absolute_paths_of_files_in_dir(shard_dir_path)
+    return batched_inferences_from_files(files_to_process, batch_size)
+
+
+def _make_tidy_df_from_seq_names_and_prediction_array(
+        sequence_names,
+        predictions_array,
+        vocab,
+        min_decision_threshold=1e-20):
+    """Given a list of sequences and a matrix of prediction values, yield a tidy dataframe of predictions."""
+    up_ids = []
+    labels = []
+    values = []
+
+    for i in range(len(sequence_names)):
+        up_id = sequence_names[i]
+        preds = predictions_array[i, :]
+
+        for vocab_index in np.argwhere(preds > min_decision_threshold):
+            vocab_index = vocab_index[0]
+            up_ids.append(up_id)
+            labels.append(vocab[vocab_index])
+            values.append(preds[vocab_index])
+    return pd.DataFrame({"up_id": up_ids, "label": labels, "value": values})
+
+
+def get_normalized_inference_results(shard_dir_path,
+                                     vocab,
+                                     label_normalizer,
+                                     min_decision_threshold=1e-20):
+    """Take a directory of sharded inferences and output a tidy and normalized dataframe.
+    
+    Args:
+        shard_dir_path: directory of TFrecord inference shards

```

Comment:
given that the predictions are stored in a custom format, it would be helpful to include a pointer to the code used to write such files. No need to document the format, just document how to write things in the format.

Reply:
Thanks done